prow/cmd/webhook-server/helpers.go

```diff
@@ -475,3 +474,33 @@ func (wa *webhookAgent) fetchClusters(d time.Duration, ctx context.Context, stat
 		}
 	}
 }
+
+func (wa *webhookAgent) fetchClustersInitially() error {
+	cfg := wa.configAgent.Config()
+	opener, err := io.NewOpener(context.Background(), wa.storage.GCSCredentialsFile, wa.storage.S3CredentialsFile)
+	if err != nil {
+		return err
+	}
+	if location := cfg.Plank.BuildClusterStatusFile; location != "" {
+		reader, err := opener.Reader(context.Background(), location)
+		if err != nil {
+			if !io.IsNotExist(err) {
+				return fmt.Errorf("error opening build cluster status file for reading: %w", err)
+			}
+			logrus.WithError(err).Warn("Build cluster status file location was specified, but could not be found. This is expected when the location is first configured, before plank creates the file.")
+		} else {
+			defer reader.Close()
+			b, err := ioutil.ReadAll(reader)
+			if err != nil {
+				return fmt.Errorf("error reading build cluster status file: %w", err)
+			}
+			var tempMap map[string]plank.ClusterStatus
+			if err := json.Unmarshal(b, &tempMap); err != nil {
+				return fmt.Errorf("error unmarshaling build cluster status file: %w", err)
+			}
+			wa.statuses = tempMap
+		}
+	}
```

Comment:
Duplicated code should be avoided. This can be simply defined as a function called something like `fetchOnce`, and in `fetchPeriodically` you can do `fetchOnce` before the for loop, then do `fetchOnce` inside the for loop

Reply:
we are doing this, because there is no assurance that the validation endpoint gets called before the clusters are initially fetched so we do need this to run serially in the main thread before the validation endpoint is reached. what i could do is put the shared logic in a function and use it in both places.

Reply:
yes my point is to avoid duplicated code, not to skip this step

Reply:
fixed